apis/r/R/SOMADataFrame.R

```diff
@@ -35,6 +44,19 @@ SOMADataFrame <- R6::R6Class(
       }
       schema <- private$validate_schema(schema, index_column_names)
 
+      stopifnot(
+        "domain must be NULL or a named list, with values being 2-element vectors or NULL" =
+          is.null(domain) || is.list(domain))
+      if (!is.null(domain)) {
+        stopifnot("domain must be NULL or a named list, with values being 2-element vectors or NULL" =
+          !is.null(names(domain)))
+        for (name in names(domain)) {
+          slot <- domain[[name]]
+          stopifnot("domain must be NULL or a named list, with values being 2-element vectors or NULL" =
+            is.null(slot) || (is.vector(slot) && length(slot) == 2))
+        }
+      }
```

Comment:
Strengthend and simplified checking of `domain`
```suggestion
      stopifnot(
        "domain must be NULL or a named list, with values being 2-element vectors or NULL" = is.null(domain) ||
          ( # Check that `domain` is a list of length `length(index_column_names)`
            # where all values are named after `index_column_names`
            # and all values are `NULL` or a two-length atomic non-factor vector
            rlang::is_list(domain, n = length(index_column_names)) &&
              identical(sort(names(domain)), sort(index_column_names)) &&
              all(vapply_lgl(
                domain,
                function(x) is.null(x) || (is.atomic(x) && !is.factor(x) && length(x) == 2L)
              ))
          )
      )
```